doc/designs/fetching-composite-algorithms.md

```diff
@@ -0,0 +1,129 @@
+# Fetching composite algorithms and using them - adding the bits still missing
+
+## Quick background
+
+We currently support - at least in the public libcrypto API - explicitly
+fetching composite algorithms (such as AES-128-CBC or HMAC-SHA256), and
+using them in most cases.  In some cases (symmetric ciphers), our providers
+also provides them.
+
+However, there is one class of algorithms where the support for *using*
+explicitly fetched algorithms is lacking: asymmetric algorithms.
+
+For a longer background and explanation, see
+[Background / tl;dr](#background-tldr) at the end of this design.
+
+## Public API - Add variants of `EVP_PKEY_CTX` initializers
+
+As far as this design is concerned, there is only class of APIs affected by
+this: DigestSign and DigestVerify.
+
+The proposal is to add these functions:
+
+``` C
+EVP_DigestSignInit_ex2(EVP_PKEY_CTX **pctx,
+                       EVP_SIGNATURE *sig, EVP_PKEY *pkey,
+                       OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
+EVP_DigestVerifyInit_ex2(EVP_PKEY_CTX **pctx,
+                         EVP_SIGNATURE *sig, EVP_PKEY *pkey,
+                         OSSL_LIB_CTX *libctx, const OSSL_PARAM params[]);
+```
+
+## Requirements on the providers
+
+Because it's not immediately obvious from a composite algorithm name what
+key type it requires / supports, at least in code, allowing the use of an
+explicitly fetched `EVP_SIGNATURE` requires that providers cooperate by
+declaring what key type is required / supported by each algorithm.
+
+There are two ways this could be implemented:
+
+1.  through an added provider function that would work like keymgmt's
+    `query_operation_name` function, but would return a key type name
+    instead:
+    
+    ``` C
+    # define OSSL_FUNC_SIGNATURE_QUERY_KEY_TYPE         26
+    OSSL_CORE_MAKE_FUNC(const char *, signature_query_key_type, (void))
+    ```
+
+2.  through a gettable `OSSL_PARAM`, using the param identity "keytype"
```

Comment:
So what are the implications for backwards compatibility of doing this? E.g. if `EVP_DigestSignInit_ex2` were added to OpenSSL 3.3, and these new provider APIs are required to use it, what happens if you call `EVP_DigestSignInit_ex2` with an `EVP_SIGNATURE` fetched from an older provider that does not supply this support?

Reply:
Good question.  I see two possible fallbacks:

1. if the fetched signature algorithm isn't a composite, but is rather something like, say, "RSA", there will probably be a corresponding keymgmt implementation with the same name.  So simply trying to fetch a keymgmt with the same name is one possibility.
2. we can fall back on existing sigalg registrations and figure out the key type from there.

But other than that, I have no idea except for "not supported".  I don't think it would be sane to try to reach the impossible, and would rather see us accepting that this new functionality depends on provider cooperation.

Reply:
That makes sense, and should probably be enough to make existing providers work with the new `EVP_DigestSignInit_ex2` function in most cases. You should probably add some discussion on this to your design document.

Reply:
Ah, one thing I forgot, which touches on this, is that I think combining a signature algorithm with a separate digest algorithm should still be allowed, as long as the providers allows it of course.  That can be passed with the OSSL_PARAM array.

Reply:
I've added a fallback strategy section